File: templates/problems/10_backtracking/03_permutations/python/solution.py
def solve_permute(nums: list[int]) -> list[list[int]]:
    """Generate all permutations using backtracking with a used array.

    Time: O(n! * n), Space: O(n) for recursion and used array
    """
    result: list[list[int]] = []
    used = [False] * len(nums)

    def backtrack(current: list[int]) -> None:
        if len(current) == len(nums):
            result.append(list(current))
            return
        for i in range(len(nums)):
            if used[i]:
                continue
            used[i] = True
            current.append(nums[i])
            backtrack(current)
            current.pop()
            used[i] = False

    backtrack([])
    return result

File: templates/problems/10_backtracking/03_permutations/typescript/solution.ts
/**
 * Generate all permutations using backtracking with a used array.
 * Time: O(n! * n), Space: O(n) for recursion and used array
 */
export function solvePermute(nums: number[]): number[][] {
    const result: number[][] = [];
    const used = new Array<boolean>(nums.length).fill(false);

    const backtrack = (current: number[]): void => {
        if (current.length === nums.length) {
            result.push([...current]);
            return;
        }
        for (let i = 0; i < nums.length; i++) {
            if (used[i]) continue;
            used[i] = true;
            current.push(nums[i]);
            backtrack(current);
            current.pop();
            used[i] = false;
        }
    };

    backtrack([]);
    return result;
}
